crates/node/src/lib.rs

```diff
@@ -323,7 +325,54 @@ pub fn run(
         .enable_all()
         .build()
         .unwrap()
-        .block_on(run_aux(config, wasm_dir, scheduled_migration));
+        .block_on(run_aux(config.ledger, wasm_dir, scheduled_migration));
+}
+
+/// Check the `tendermint_mode` has changed from validator to non-validator
+/// mode, in which case we replace and backup the validator keys and state to
+/// avoid CometBFT running as a validator. We also persist the
+/// `last_tendermint_node` in the config for the next run.
+fn handle_tendermint_mode_change(config: &config::Config) {
+    // Check if the node was previously ran as a Validator, but isn't anymore
+    if !matches!(
+        config.ledger.shell.tendermint_mode,
+        TendermintMode::Validator
+    ) && matches!(
+        config.ledger.shell.last_tendermint_mode,
+        Some(TendermintMode::Validator)
+    ) {
+        // Backup and replace CometBFT validator key and state
+        let cometbft_dir = config.ledger.cometbft_dir();
+        namada_apps_lib::tendermint_node::backup_validator_key_and_state(
+            &cometbft_dir,
+        );
+        namada_apps_lib::tendermint_node::write_dummy_validator_key_and_state(
```

Comment:
It looks to me like we could enter this branch even if the node is transitioning from a full node to a validator, is it ok to write this dummy data in this case?

Reply:
I think the second part of the condition should be `false` in that case (`last_tendermint_mode == Full`)

Reply:
My bad, I didn't see the second `matches!`, thought there was only one